ledger/acctupdates.go

```diff
@@ -350,181 +352,216 @@ func (au *accountUpdates) ListApplications(maxAppIdx basics.AppIndex, maxResults
+			}
 		}
-	}
-	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
 
-	// Check for creatables that haven't been synced to disk yet.
-	var unsyncedCreatables []basics.CreatableLocator
-	deletedCreatables := make(map[basics.CreatableIndex]bool)
-	for _, cidx := range keys {
-		delta := au.creatables[cidx]
-		if delta.created {
-			// Created but only exists in memory
-			unsyncedCreatables = append(unsyncedCreatables, basics.CreatableLocator{
-				Type:    delta.ctype,
-				Index:   cidx,
-				Creator: delta.creator,
-			})
-		} else {
-			// Mark deleted creatables for exclusion from the results set
-			deletedCreatables[cidx] = true
+		// Check in-memory created creatables, which will always be newer than anything
+		// in the database
+		var res []basics.CreatableLocator
+		for _, loc := range unsyncedCreatables {
+			if uint64(len(res)) == maxResults {
+				au.accountsMu.RUnlock()
+				return res, nil
+			}
+			res = append(res, loc)
 		}
-	}
 
-	// Check in-memory created creatables, which will always be newer than anything
-	// in the database
-	var res []basics.CreatableLocator
-	for _, loc := range unsyncedCreatables {
-		if uint64(len(res)) == maxResults {
-			return res, nil
+		au.accountsMu.RUnlock()
+
+		// Fetch up to maxResults - len(res) + len(deletedCreatables) from the database,
+		// so we have enough extras in case creatables were deleted
+		numToFetch := maxResults - uint64(len(res)) + uint64(len(deletedCreatables))
+		dbResults, dbRound, err := au.accountsq.listCreatables(maxCreatableIdx, numToFetch, ctype)
+		if err != nil {
+			return nil, err
 		}
-		res = append(res, loc)
-	}
 
-	// Fetch up to maxResults - len(res) + len(deletedCreatables) from the database,
-	// so we have enough extras in case creatables were deleted
-	numToFetch := maxResults - uint64(len(res)) + uint64(len(deletedCreatables))
-	dbResults, err := au.accountsq.listCreatables(maxCreatableIdx, numToFetch, ctype)
-	if err != nil {
-		return nil, err
-	}
+		if dbRound == currentDbRound {
+			// Now we merge the database results with the in-memory results
+			for _, loc := range dbResults {
+				// Check if we have enough results
+				if uint64(len(res)) == maxResults {
+					return res, nil
+				}
+
+				// Creatable was deleted
+				if _, ok := deletedCreatables[loc.Index]; ok {
+					continue
+				}
 
-	// Now we merge the database results with the in-memory results
-	for _, loc := range dbResults {
-		// Check if we have enough results
-		if uint64(len(res)) == maxResults {
+				// We're OK to include this result
+				res = append(res, loc)
+			}
 			return res, nil
 		}
-
-		// Creatable was deleted
-		if _, ok := deletedCreatables[loc.Index]; ok {
-			continue
+		if dbRound < currentDbRound {
+			au.log.Errorf("listCreatables: database round %d is behind in-memory round %d", dbRound, currentDbRound)
+		}
+		au.accountsMu.RLock()
+		for currentDbRound >= au.dbRound && currentDeltaLen == len(au.deltas) {
```

Comment:
it's not. The `currentDbRound` is the value of `au.dbRound` on the last time we took the lock and evaluated the memory content.

What I was looking for is to have a change in the `au.dbRound`. Since `au.dbRound` can move only forward, the normal condition is that `currentDbRound < au.dbRound`. however, since we're blocking in this for-loop, I need to take the complement of that expression.
`!(currentDbRound < au.dbRound)` -> `currentDbRound >=  au.dbRound`